### ISSUE TITLE
Grouping by Date Parts

### ISSUE THREAD

**Issue by TehWardy** [dotnet/efcore#22008]
**Reproducing the problem**

```cs
var results = Db.GetAll<Invoice>()
    .GroupBy(t => $"{t.DueDate.Year}-{t.DueDate.Month}")  <--- using t.Due.ToString("yyyy-MM") also doesn't work here.
    .Select(g => new TotalsByDueDate<Invoice>
        { 
        CalendarMonth = g.Key,
        Count   = g.Count(),
        Sum     = g.Sum(i => i.Value),
        Average = g.Average(i => i.Value),
        Min     = g.Min(i => i.Value),
        Max     = g.Max(i => i.Value),
        Data    = g
    })
    .ToArray();
```

EF Core 3.1.6 seems to be unable to translate this "seemingly simple" scenario to a DB query.

**In short**
Is this scenario supported?

If not ... 
How do I resolve this without having to pull all the date from the DB before the group by? 
Required as it would result in a lot of data needing to be pulled from the db.

**Further details** 

Captured output from aspnet logging on the OData controller used to "build the query" ...

```
2020-08-10 17:41:45,076 [12] ERROR  Api.IAppBuilderExtensions - The LINQ expression 'DbSet<Invoice>
    .Where(i => ... )  <-- middle of query cut for berevity (where clause injected by query filter on the set)
    .GroupBy(
        source: i => string.Format(
            format: "{0}-{1}", 
            arg0: (object)i.DueDate.Year, 
            arg1: (object)i.DueDate.Month), 
        keySelector: i => i)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
```

I don't think this is an "un-translatable scenario".

**Comment by smitpatel** [comment]
- Interpolated strings are not translated. You can use `+` operator to combine strings.
- DateTime.ToString with format cannot be translated to server since it requires manipulating database datetime and database would have different semantics than .NET
- Your Select after GroupBy contains reference to g full `data = g`. which means you want whole group. We don't support GroupBy which gets groupings back since there is no SQL equivalent of IGrouping. See https://github.com/dotnet/efcore/issues/17068#issuecomment-586464350 for more details. You are anyway getting whole data back to client by get all groupings so if you want groups then put AsEnumerable before GroupBy and apply GroupBy on client side.

**Comment by TehWardy** [comment]
My understanding is that by referencing the "set" referred to by the group I would only pull that set if I issued an include to query.
This is included in the result set in order that if an OData expand is provided in the query URL then the data is returned but if it isn't part of the OData url then it should not be loaded.

If I exclude the set is it at all possible to make this grouping expression work as using the + operator on the year and month from a date would just sum two integers.

In short ...
How do I group sets by their "calendar month"?

**Comment by smitpatel** [comment]
That is not how "set" of groups work. You should look up LINQ GroupBy operator.

`+` is also defined for string. Just call `ToString` before combining so that it does string concat rather than int sum.

**Comment by TehWardy** [comment]
So this is where my confusion comes in because EF Core clearly doesn't support "group by" in the sense that IEnumerable<T>.GroupBy() returns an IEnumerable<IGrouping<TKey, TElement>> and EF Core doesn't seem to support IGroupings.
This alone implies that EF Core doesn't support grouping full stop.

My understanding from what I've read on here and the documentation for EF Core's "Complex Operators" ...
     https://docs.microsoft.com/en-us/ef/core/querying/complex-query-operators
... is that there's some breaking point where code cannot be mapped to a relevant SQL expression.

I'm just not clear what that breaking point is because that clearly shows that grouping IS supported.

My understanding is that given OData's limitations both frameworks should "mostly share" a logical limit for the kinds of grouping that can be done.

In this case I need a date formatted as a particular string from "some of it's parts" as my group key which is obviously a thing that can be done in SQL, LINQ and in OData, my issue here is how to tell EF this since EF lately particular about what it will and won't accept.

So ...
Exactly how do I tell EF Core ...

```
 .GroupBy(i => i.Date.ToString("yyyy-MM"))
```

... in a way that it will accept?

As a secondary question to this what is the rule for grouping in general and when will EF support OData Apply / aggregation operations again (as this used to be fine in EF6)?

**Comment by smitpatel** [comment]
https://github.com/dotnet/efcore/issues/17068#issuecomment-586464350 describes in the detail what is supported, what will be supported and what will never be supported.
LINQ GroupBy is not same as SQL GROUP BY clause. There will be limitations and you cannot compare both of them. Further, it is not just GroupBy operator, there is whole combination of operators working together which determines if a pattern is supported or not. There is no way to detail it in writing which covers all supported scenario. Hence above link specify big bracket and up to the user to understand if it matches their own queries.
Bottom line, if you have a group by query in LINQ and you have a SQL translation which will get the same result, we can add support for it. If you cannot write an equivalent SQL for your LINQ query then take it is not supported scenario.

DateTime.ToString("yyyy-MM") does not have direct server equivalent so it is not translated to server and any operator which requires its translation will fail. (like GroupBy above but same expression inside Where will also fail).

**Comment by TehWardy** [comment]
I didn't see anything that mentioned formatted / translated date strings in that, although scalar values are supposedly already supported and documented and this is a scalar grouping scenario (of sorts).

For this ticket I'm just looking to figure out what the EF team recommends we do for the breaking of Group By as consumed by OData Apply and aggregation operations for years ... in short "is there a method for mapping this kind of thing"?

I agree, the fact that the two group by operations aren't the same causes some incompatibility between EF and OData which can't be resolved in the current form but that's down to certain design choices in the query builder I guess. 

Taking an objective "outsiders look" at this though ...

From the documentation on this particular subject here:
   https://docs.microsoft.com/en-us/ef/core/querying/complex-query-operators

> LINQ GroupBy operators create a result of type `IGrouping<TKey, TElement>` where `TKey `and `TElement `could be any arbitrary type. Furthermore, `IGrouping `implements `IEnumerable<TElement>`, which means you can compose over it using any LINQ operator after the grouping. Since no database structure can represent an `IGrouping`, GroupBy operators have no translation in most cases. When an aggregate operator is applied to each group, which returns a scalar, it can be translated to SQL GROUP BY in relational databases. The SQL GROUP BY is restrictive too. It requires you to group only by scalar values. The projection can only contain grouping key columns or any aggregate applied over a column. 

I don't believe this is correct because ...
 - The igrouping could be represented as a "selected" flat set with an additional set of columns for the group added to that set.
      - this means NOT using sql group by at all and relying on the entity mapper to handle it instead when the query is loaded from the SQL server to the application.
 - Further compositions could then just be applied to that flat set on either side by performing a selection on that query.
 - Keys could be computed as long as the linq query contains simple "sql translatable operations" within the key building functions.

It boils down to "how" the LINQ query is mapped.
My approach allows for ...
  - More complex keys
  - returning the set of sets (so to speak)
  - composition on a groupby result.

Or am I missing something here?

I'm still not sure you answered my original question though ...
How do I group by calendar month?

**Comment by smitpatel** [comment]
`GroupBy(t => t.DueDate.Year.ToString() + "-" + t.DueDate.Month.ToString())`

No your question is not GroupBy, your question is about Calendar month. Just because it demonstrate while you are using GroupBy it has nothing to do with limitations of GroupBy. I specified above also that same thing even if used in Where will also fail. So I don't see any reason of keep discussing GroupBy limitations.

After extensively working on GroupBy LINQ queries since EF Core 2.1 things are easier said than done because the complex expression tree compiler gives you. Even if you have a valid SQL translation, connecting it with LINQ query is no easy task. Though I am not saying everything is impossible. Things can be possible and we may have overlooked.

So if you have a LINQ query in mind and know equivalent SQL translation, let us know, we will try to support it.

If you think you have fundamental approach on how to translate GroupBy differently than how EF Core treating it, then please submit a design proposal with different expression trees generated, how you would identify patterns, process the tree if needed, translate it to server and finally generate results out of it. If you want to go an extra kilometer you can also submit a PR, we will be glad to accept a well-written and tested PR.

**Comment by sdanyliv** [comment]
> How do I group by calendar month?

It will not solve problem that you have with selecting `g` in projection. But if EF Core supports DatePart translation on the fly:
```cs
var results = Db.GetAll<Invoice>()
    .GroupBy(t =>  new { t.DueDate.Year, t.DueDate.Month })
    .Select(g => new TotalsByDueDate<Invoice>
        { 
        CalendarMonth = g.Key,
        Count   = g.Count(),
        Sum     = g.Sum(i => i.Value),
        Average = g.Average(i => i.Value),
        Min     = g.Min(i => i.Value),
        Max     = g.Max(i => i.Value),
        Data    = g // remove that, it is not translatable and has no sense for Sql Server grouping
    })
    .ToArray();
```
If you still need that property, you have to do client side grouping 
```cs
var results = Db.GetAll<Invoice>()
    .AsEnumerable()
    .GroupBy(t =>  new { t.DueDate.Year, t.DueDate.Month })
    .Select(g => new TotalsByDueDate<Invoice>
        { 
        CalendarMonth = g.Key,
        Count   = g.Count(),
        Sum     = g.Sum(i => i.Value),
        Average = g.Average(i => i.Value),
        Min     = g.Min(i => i.Value),
        Max     = g.Max(i => i.Value),
        Data    = g
    })
    .ToArray();
```

**Comment by TehWardy** [comment]
Oh ... that's interesting ... 
I was under the impression that "non scalar" or "non column value" based grouping wasn't possible at all for the same reason @smitpatel highlighted above about SQL having "no direct translation".
The construction of an annon object here I thought was well out of bounds as that was where my initial thoughts went due to it "keeping the type safety" but i didn't write it this way thinking it wouldn't work.

Maybe i'm missing something about how the underlying query translation works ... 
I'll have to have a read through the code but last time I checked It was extremely "expression tree building heavy" and that stuff hurts my head as it's not your typical C# (for most people).

**Comment by sdanyliv** [comment]
> I'll have to have a read through the code but last time I checked It was extremely "expression tree building heavy" and that stuff hurts my head as it's not your typical C# (for most people).

Analyzing EF Core translator is not a trivial task as a LINQ translation itself. But you had to google: LINQ how to group by several fields - obvious question.

**Comment by TehWardy** [comment]
Right but my issue isn't "how to do this in LINQ" it's "how to do it with LINQ to entities" which is a considerably more limited scenario in .Net Core specifically. 

In EF6 under .Net 4 this stuff was a ton more "flexible" (in that supported more grouping scenarios) ... that said I agree with the dev teams reasoning for it not working that way any more as the query translator has had miracles performed on it since EF6 under .Net 4.

This boils down to me not knowing where the line is drawn with the current query translator in terms of "what can it do".
Complex grouping criteria (such as the usage of an anon object) I was under the impression was not supported.

In short: @sdanyliv it's not quite as obvious as it seems on face value.

**Comment by roji** [comment]
It's indeed not always trivial to know what EF Core can translate and what it can't - we've been investing quite a bit of effort in improving translation failure messages for 5.0, so that they point out exactly what part of the LINQ query is problematic. I think that's the best way to improve people's understanding of what can be translated and what cannot.

As a general rule, some functions/constructs simply have no corresponding construct in SQL; interpolated strings is one good example. The translation failure exception says that `string.Format` cannot be translated (that's what interpolated strings get compiled into). The reason is that there simply isn't a SQL function that corresponds to C# string.Format, with the interpolation and formatting features etc. String concatenation, on the other hand, is a much simpler concept, and SQL obviously has concatenation support as well.

Anonymous objects are frequently a "temporary artifact" inside your query; they function as a way for you to express what you want to EF Core's query pipeline. In @sdanyliv's example above, the temporary object is used to tell EF to group by two expressions, and so the final query contains `GROUP BY DATEPART(year, [i].[DueDate]), DATEPART(month, [i].[DueDate])`. Note that there isn't actually an anonymous object in the LINQ query (that concept does not exist) - the EF query pipeline translates the anonymous object to the GROUP BY clause.

Finally, LINQ's GroupBy is indeed a complicated operator, because in some case it corresponds to SQL GROUP BY, but in others it does not. [As our docs on GroupBy say](https://docs.microsoft.com/en-us/ef/core/querying/complex-query-operators#groupby), SQL's GROUP BY can only be used with an aggregate function (count, max...) which reduces all the grouped rows's columns into scalar values. This means that any usage of LINQ GroupBy which doesn't reduce the outputting groups is simply non-translatable to SQL - this isn't an EF limitation, but a "mismatch" between what GroupBy means in LINQ and in SQL.

@TehWardy I hope the above helps understand the (different) reasons why your original query cannot be translated, and why @sdanyliv's [suggested query](https://github.com/dotnet/efcore/issues/22008#issuecomment-672714856) does. I'll go ahead and close this as there doesn't seem to be anything actionable on EF Core's side, but please feel free to continue asking here, and we can always reopen if necessary.

**Comment by TehWardy** [comment]
@roji  that's awesome, thanks for the input. It's definitely a learning curve that the compiler / translation handler can help with.

Tickets like this I think act as a guide for you guys to help understand the frustration we as consumers have when something on face value is valid LINQ but not valid LINQ to Entities (for a weird reason like a technology gap).

When I read ...
> DateTime.ToString("yyyy-MM") does not have direct server equivalent

I was thinking "that's nuts", SQL can of course take parts of dates, build a string from those and then group on them. 
It hadn't occurred to me though the reason I had to take a different path or the underlying problem that a .ToString("format") call specifically presents.

One thing i'm not sure of ...
Is there a path to OData's aggregation functions still open here?

The OData framework has support for them but when I try to use them I get errors that read "maybe due to some limitation in EF". 
Generally speaking I'm looking to build graphs or charts on web pages by constructing aggregate ranges over time periods as a sort of pattern.
Having gotten the query working from the above I could just build custom functions on the controller but it would be a much cleaner solution if I could do the same thing as an OData query.

**Comment by roji** [comment]
I'm personally not very familiar with OData or the query shapes it represents, but we are actively working on better test coverage for OData for 5.0, and definitely strive to support common query shapes. The best way forward with these kinds of issues, is to post the LINQ query generated by OData and going into EF Core, along with the OData query that produced it - that provides the most information for us to analyze.

>> DateTime.ToString("yyyy-MM") does not have direct server equivalent
>
> I was thinking "that's nuts", SQL can of course take parts of dates, build a string from those and then group on them. [...]

Of course you're right - SQL Server does have ways of taking parts of dates and building strings from them, but not exactly in the same way as DateTime.ToString. In other words, I'm not aware of a SQL function that can accept `yyyy-MM` (as well as the other codes and formats supported by DateTime.ToString). This is why you must use other means of extracting and concatenating in LINQ, which can be translated.

**Comment by TehWardy** [comment]
Yeh ... summed it up perfectly!
It's obvious now I think about it, as always though missing / not thinking about what a framework does with the instructions you give it can be an issue.

Maybe you guys can work with @xuzhg to get the usages of odata "apply" and "aggregate" synched up.
I'm sure the changes needed are fairly small but this might be a a similar issue in that not all OData query expressions are directly translatable to SQL when you get in to the details.

My gut feeling is that most of the breakages are much like this ticket.
Next time I hit such a break i'll see if I can extract the expression tree info and post it here (as a new ticket).
Worst case the advice you guys give me is that the OData query translator needs to build slightly different LINQ expressions (gut feeling).

It would be nice to get the gaps closed out between these two technologies.

**Comment by sdanyliv** [comment]
@TehWardy, Offtop but why not,
I think i have fixed almost all issues with OData in linq2db since they produce very interesting (clever) projection. So if EF Core still can not translate OData’s query, just try this EF Core extension [linq2db.EntityFrameworkCore](https://github.com/linq2db/linq2db.EntityFrameworkCore)

**Comment by TehWardy** [comment]
Oh that's interesting, I didn't know that existed.
Is LinqToDb an alternaitve implementation of LinqToEntities or is this a means to extend the core functionality?

If this is an extension to support the missing functionality how come you didn't do what smitpatel suggested above and commit a pull request to get this in to the core framework?

Typically I have avoided taking on things like this as it's then not always clear if this problem is in Microsoft code or in some extension and my requirements whilst they do have some edge cases I would like to see working (like this one) generally my problems can be worked around using simpler features already implemented.
OData queries however are an interesting case as I believe it was ajvickers who said once "produces some interesting queries". 

As always my first goal is to have Microsoft teams work on their gaps between frameworks which I think they are generally good with the idea if you can show them the gap, the issue is usually getting the framework to show me the information I need in order to report a meaningful ticket to one team or another.

Have you considered committing a pull request to the EFCore team to pull in some of the edge cases they currently don't support from your codebase?

**Comment by sdanyliv** [comment]
Linq2db is an alternative lightweight ORM without change tracking. We just reused EF Core mapping, but LINQ query is parsed by linq2db itself. So no PR here ;)
When people can not create SQL they move to Dapper - it is not our case, we can express almost any SQL via LINQ, that’s why this alternative sill exists and popular among people who know SQL and needs highly optimized queries.

**Comment by TehWardy** [comment]
Can this replace EF?
I'd still keep it for the migrations and what not but if it works as you say there's a chance it can perform a more complete backing store implementation for OData.

"use the right tool for the job" comes to mind.
One thing I swear by in EF is the migrations.

**Comment by sdanyliv** [comment]
Just try `ToLinqToDB()` and compare SQL.

**Comment by TehWardy** [comment]
I rarely actually build the queries directly I rely on OData to do it for me from the URL.
If I install this extension you linked to above does it trigger OData to do that automatically or do I have to wire something up?

**Comment by sdanyliv** [comment]
Post sample of the method, i’ll figure out how to do that.

**Comment by TehWardy** [comment]
**Disclaimer**
> I don't what is the break here but I know what i'm trying to achieve.
> There's a ton of moving parts and it's in the connection of multiple frameworks that I have my issue.
> Let me know if this is not clear.

**OData as Documented**
It's tricky but the general idea is that I have an entity set in a "standard EF Core DbContext" with an applied filter on it which I compute based on a fixed rule for the current user (determined using a session token) (just applies a where clause to the base set the user can pull from the db).

Then the OData controller (My base controller for all sets `EntityController<T>`) basically calls in to that like this ...

```
[HttpGet]
[EnableQuery(
    AllowedArithmeticOperators = AllowedArithmeticOperators.All,
    AllowedFunctions = AllowedFunctions.AllFunctions,
    AllowedLogicalOperators = AllowedLogicalOperators.All,
    AllowedQueryOptions = AllowedQueryOptions.All,
    MaxAnyAllExpressionDepth = 3,
    MaxExpansionDepth = 3
)]
public virtual IActionResult Get(ODataQueryOptions<T> queryOptions) => Ok(Db.GetAll<T>());
```

The "queryOptions" are applied "after my code has finished executing and returned" by the framework prior to serialising the result set.

**Moving on to the problem**
The idea is that the query is supposed to build out a groupby then provide an array of "group totals" with each aggregation providing the property values.

So "in theory" something like this in an OData query form would be the same as my query from above (the example you asked for):
  (NOTE: i've added some white space here to help with parsing the query but this would be a URL)

```
/Invoice?$apply=groupby(
    (year(DateTime), month(DateTime)),
     aggregate(Id with countdistinct as Count),
     aggregate(Value with sum as Sum),
     aggregate(Value with average as Average),
     aggregate(Value with min as Min),
     aggregate(Value with max as Max)
)

```
This follows the pattern shown here:
   https://devblogs.microsoft.com/odata/aggregation-extensions-in-odata-asp-net-core/

This should be translated to a valid LINQ expression tree that then in turn would be translated to a valid SQL query (the root of my frustration) .
This is ultimately what I came here to try and solve but didn't want to bog down the EF Core team with my OData woes by presenting this as a "possible Odata problem", so I hit this by making a new custom action method on the controller and implementing the logic myself (resulting in the above query).

The issues ...
 - Proving that EF Core can do this. (check)
 - Proving that OData can do this. (check)
      - https://docs.microsoft.com/en-us/odata/changelog/webapi-7x
      - Which links to this: https://github.com/OData/WebApi/pull/1728 (which I thought solved what i need)
      - I added a comment to the bottom there which contains an exception pointing at EF Core
 - Proving they can do this together. (the sticky bit)

In practice I've never been able to get this working under .Net Core despite having some success a while back with OData 3 and .Net 4.

If someone solves this they have a friend for life!
This will basically beat all my existing pain points with complex queries in one hit.

**Comment by sdanyliv** [comment]
I assume GetAll is IQueryable
```cs
=> Ok(Db.GetAll<T>().ToLinqToDB());
```
